### PR TITLE
Fix label ordering

### PR DESCRIPTION
**Pull request by balderdash**
Trailing labels aren't being rendered correctly under the current custom block parser, e.g.:
```
  {
    "func": "blahblah",
    "blockText": "before input {INPUT} after input",
    "args": [
      {
        "name": "INPUT"
      }
    ]
  }
```
Before:
image
After:
image

Things get really weird with multi-line blocks:
```
  {
    "func": "dosomething",
    "blockText": "val 1 {VALUE1} field input {FIELD} val 2 {VALUE2} end",
    "args": [
      {
        "name": "FIELD",
        "field": true
      },
      {
        "name": "VALUE1"
      },
      {
        "name": "VALUE2"
      }
    ],
    "inline": false
  }
```
Before:
image

After:
image

The code change to make this happen was grouping inputs into `inputRows` and handling each one independently (replacing the broken and ugly `getPreviousInput` approach I had tried earlier). Each `inputRow` is just a group of inputs ending with a line-ending input, i.e. a value, statement, or dummy input. The labels and field inputs in each row can now be assembled in a fairly straightforward way:
https://github.com/code-dot-org/code-dot-org/blob/33d6e59341a0cf3409ed122878b54f18e88ed4b7/apps/src/block_utils.js#L577-L590

For inline blocks, the rows all just string together in a long line, looking just like `blockText` (but a block),

The other bit of refactoring that made sense to do here was adopting the `customInputTypes` config format for all the input types (this is the thing I added for the location picker input in https://github.com/code-dot-org/code-dot-org/pull/21955).

### CODE REVIEW

File: apps/src/block_utils.js [hunk 1]
```diff
@@ -485,63 +486,108 @@ const determineInputs = function (text, args, strictTypes=[]) {
...
 /**
  * Adds the specified inputs to the block
  * @param {Blockly} blockly The Blockly object provided to install()
  * @param {Block} block The block to add the inputs to
  * @param {Object[]} inputs The list of inputs. See determineInputs() for
  *   the fields in each input.
- * @param {object} customInputTypes A map of customType input names to their
- *   definitions, which are objects that have `addInput` and `generateCode`
+ * @param {object} inputTypes A map of input types to their definitions, which
+ *   are objects that have `addInput` and `generateCode`
```
Comment (Hamms) [inline apps/src/block_utils.js]: nit: could we get a jsdoc `@typedef` for input definitions?

File: apps/src/block_utils.js [hunk 2]
```diff
@@ -485,63 +486,108 @@ const determineInputs = function (text, args, strictTypes=[]) {
...
+  inputRows.push(lastGroup);
+  inputs.forEach(input => {
+    lastGroup.push(input);
+    if (inputTypes[input.mode].addInputRow) {
+      lastGroup = [];
+      inputRows.push(lastGroup);
     }
   });
+  const lastRow = inputRows[inputRows.length - 1];
+  if (inputRows[inputRows.length - 1].length > 0) {
```
Comment (Hamms) [inline apps/src/block_utils.js]: nit: could just be `lastRow.length`

File: apps/src/block_utils.js [hunk 3]
```diff
@@ -701,24 +752,8 @@ exports.createJsWrapperBlockCreator = function (
...
-          }
-          return code;
-        } else if (arg.statement) {
-          const code = Blockly.JavaScript.statementToCode(this, arg.name);
-          return `function () {\n${code}}`;
-        } else {
-          return Blockly.JavaScript.valueToCode(this, arg.name, ORDER_COMMA);
-        }
+        const inputConfig = inputConfigs.find(input => input.name === arg.name);
+        return inputTypes[inputConfig.mode].generateCode(this, inputConfig);
```
Comment (Hamms) [inline apps/src/block_utils.js:782]: :+1:

**Comment by balderdash** [comment]
There are some (clearly insufficient) unit tests in `blockUtilsTest.js`, but I'd like to add some kind of block screenshot tests that actually verify that things are rendering the way we want them to.

**Pull request merged**